client/src/unifyfs-fixed.c

```diff
@@ -270,84 +120,100 @@ void unifyfs_rewrite_index_from_seg_tree(void)
+        indexes[idx].file_pos = node->start;
+        indexes[idx].log_pos  = node->ptr;
+        indexes[idx].length   = node->end - node->start + 1;
+        indexes[idx].gfid     = gfid;
+        idx++;
     }
-
-    /* reset our segment count since we just dumped them all */
-    unifyfs_segment_count = 0;
+    seg_tree_unlock(&meta->extents_sync);
+    /* All done processing this files writes.  Clear its seg_tree */
+    seg_tree_clear(&meta->extents_sync);
 
     /* record total number of entries in index buffer */
     *unifyfs_indices.ptr_num_entries = idx;
 }
 
 /*
- * Sync all the extents to the server.  Clears the metadata index afterwards.
+ * Sync all the write extents for the target file(s) to the server.
+ * The target_fid identifies a specific file, or all files (-1).
+ * Clears the metadata index afterwards.
  *
  * Returns 0 on success, nonzero otherwise.
  */
-int unifyfs_sync(void)
+int unifyfs_sync(int target_fid)
 {
-    /* NOTE: we currently ignore gfid and sync extents for all files in
-     * the index. If we ever switch to storing extents in a per-file index,
-     * we can support per-file sync. */
-
-    /* write contents from segment tree to index buffer
-     * if we're using that optimization */
-    if (unifyfs_flatten_writes) {
-        unifyfs_rewrite_index_from_seg_tree();
-    }
+    int ret = UNIFYFS_SUCCESS;
 
-    /* if there are no index entries, we've got nothing to sync */
-    if (*unifyfs_indices.ptr_num_entries == 0) {
-        return UNIFYFS_SUCCESS;
+    if (target_fid == -1) {
+        /* ensure any data written to the spill over file is flushed */
+        ret = unifyfs_logio_sync(logio_ctx);
+        if (ret != UNIFYFS_SUCCESS) {
+            LOGERR("failed to sync logio data");
+        }
     }
 
-    /* ensure any data written to the spill over file is flushed */
-    int ret = unifyfs_logio_sync(logio_ctx);
-    if (ret != UNIFYFS_SUCCESS) {
-        LOGERR("failed to sync logio data");
-        return EIO;
-    }
+    /* For each open file descriptor .. */
+    for (int i = 0; i < UNIFYFS_MAX_FILEDESCS; i++) {
+        /* get file id for each file descriptor */
+        int fid = unifyfs_fds[i].fid;
+        if (-1 == fid) {
+            /* file descriptor is not currently in use */
+            continue;
+        }
 
-    /* tell the server to grab our new extents */
-    ret = invoke_client_sync_rpc();
-    if (ret != UNIFYFS_SUCCESS) {
-        /* something went wrong when trying to flush key/values */
-        LOGERR("failed to flush key/value index to server");
-        return EIO;
-    }
+        /* is this the target file? */
+        if ((target_fid != -1) && (fid != target_fid)) {
+            continue;
+        }
 
-    /* flushed, clear buffer and refresh number of entries
-     * and number remaining */
-    clear_index();
+        unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(fid);
+        if ((NULL == meta) || (meta->fid != fid)) {
```

Comment:
`(meta->fid != fid)` <- is this check needed?

Reply:
When fds/fids are re-used, the fid field is reset to -1. This is just a paranoid check to make sure the meta is for the current fid.

Reply:
I wanted to put the check in the `unifyfs_get_meta_from_fid()` function itself, but it's used to do get the filemeta pointer when a fid is first allocated (before fid is set).

Reply:
Got it, thanks for the info